### PR TITLE
DSR 3.0 Improve External DB Connection Management

### PR DESCRIPTION
**Pull request by pattisdr**
Closes #PROD-1881

❗ Dependent on https://github.com/ethyca/fides/pull/4760

### Description Of Changes

Improve external db connection management

### Code Changes

* [ ] Fix a regression in DSR 2.0 and apply same fix to DSR 3.0 that closes a connection after it's been used by using TaskResources as a context manager

### Steps to Confirm

- In your toml file, toggle DSR 3.0 to on and task_always_eager to True (to test celery)
 ```
[execution]
use_dsr_3_0 = true
```
```
[celery]
task_always_eager = true
```

-  nox -s dev -- shell postgres mongodb worker
- In another terminal, cd /fides/clients
 - turbo run dev
-  Setup postgres connector
- http://localhost:3001/add-systems
- Click Add a System
- Add system name: `postgres_example_test_dataset`
- Click Integrations
- Integration type: PostgreSQL
- Add credentials for local postgres example container running in docker (these aren't private secrets, these are for the example postgres database, docker/docker-compose.integration-postgres.yml). Test connection
image
-  Upload postgres_example_test_dataset.yml and click Save
- In a separate terminal, start watching the number of connections in your postgres external database:
```bash
PGPASSWORD=postgres watch 'psql -h localhost -p 6432 -U postgres -d postgres_example -c "SELECT sum(numbackends) FROM pg_stat_database"'
```
- Rapidly make this request in a separate terminal - monitored connections should stay around 2-3, instead of quickly maxing out when it gets to 100 
```bash
- curl -X 'POST' \
  'http://localhost:8080/api/v1/privacy-request' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '[
  {
    "identity": {
      "email": "customer-1@example.com"
    },
    "policy_key": "default_access_policy"
  }
]'
```

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

**Comment by pattisdr** [comment]
@galvana if you had a minute, looking for a quick review on this work that closes the connections to external databases for both DSR 2.0 and DSR 3.0.

File: src/fides/api/task/deprecated_graph_task.py [hunk 1]
```diff
@@ -109,54 +109,54 @@ def run_access_request_deprecated(
 ) -> Dict[str, List[Row]]:
     """Deprecated: Run the access request sequentially in-memory using Dask"""
     traversal: Traversal = Traversal(graph, identity)
-    resources = TaskResources(
+    with TaskResources(
         privacy_request, policy, connection_configs, EMPTY_REQUEST_TASK, session
```
Comment (pattisdr) [inline src/fides/api/task/deprecated_graph_task.py:113]: This was the DSR 2.0 regression I caused, using this a context manager had functionality to close open connections

File: src/fides/api/task/execute_request_tasks.py [hunk 2]
```diff
@@ -227,23 +227,32 @@ def run_access_node(
...
-            ]
-            # Run the main access function
-            graph_task.access_request(*upstream_access_data)
-            log_task_complete(request_task)
+            with TaskResources(
+                privacy_request,
+                privacy_request.policy,
+                session.query(ConnectionConfig).all(),
+                request_task,
+                session,
```
Comment (pattisdr) [inline src/fides/api/task/execute_request_tasks.py:235]: This mimics what DSR 2.0 is doing so we're closing connections to external databases after the request is done

File: src/fides/api/task/task_resources.py [hunk 3]
```diff
@@ -166,3 +169,14 @@ def get_all_cached_erasures(self) -> Dict[str, int]:
...
+    def get_connector(self, key: FidesKey) -> Any:
+        """Create or return the client corresponding to the given ConnectionConfig key"""
+        if key in self.connection_configs:
+            return self.connections.get_connector(self.connection_configs[key])
+        raise ConnectorNotFoundException(f"No available connector for {key}")
+
+    def close(self) -> None:
+        """Close any held resources"""
+        logger.debug("Closing all task resources for {}", self.request.id)
+        self.connections.close()
```
Comment (pattisdr) [inline src/fides/api/task/task_resources.py:182]: I had also removed this originally when I removed DSR 2.0 altogether in the POC - this closes the open connections

**Review by galvana — APPROVED** [review]
Good old `watch` command! I verified this by monitoring connections on the `dsr_parallelization` branch and this branch. Confirmed the connections don't stay open for each privacy request ✅

**Pull request merged**